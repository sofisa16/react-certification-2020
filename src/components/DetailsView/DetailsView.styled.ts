import styled from 'styled-components';
import Typography from '@material-ui/core/Typography';

const ParentContainer = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: row;
  padding: 20px;
  box-sizing: border-box;
`;

const RightContainer = styled.div`
  width: 70%;
  height: 100%;
  position: relative;
  display: flex;
  flex-direction: column;
  padding-right: 15px;
`;

const LeftContainer = styled.div`
  width: 30%;
  height: 100%;
  overflow: scroll;
`;

const DownContainer = styled.div`
  padding: 15px 30px;
`;

const TitleContainer = styled.div`
  padding-bottom: 15px;
  display: grid;
  grid-template-columns: 1fr auto;
`;

const JustifyText = styled(Typography)`
  text-align: justify;
`;

const Iframe = styled.div`
  > iframe {
    width: 100%;
    height: 500px;
  }
`;

const Styled = {
  ParentContainer,
  RightContainer,
  LeftContainer,
  DownContainer,
  TitleContainer,
  JustifyText,
  Iframe,
};

export default Styled;