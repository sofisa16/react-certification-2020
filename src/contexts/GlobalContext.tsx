import React, {useState, useEffect, useReducer} from 'react';
import {ThemeProvider} from 'styled-components';
import NoSsr from '@material-ui/core/NoSsr';
import {
  createMuiTheme,
  ThemeProvider as MuiThemeProvider,
  Theme,
} from '@material-ui/core/styles';
import {initialState, reducer, FavoriteVideos, Action} from './../components/FavoritesView/addToFavorites';
import {AUTH_STORAGE_KEY, AUTH_AVATAR, FAVORITES_VIDEOS} from './../utils/constants';
import {storage} from './../utils/storage';
import { useAuth0 } from "@auth0/auth0-react";

interface GlobalContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
  darkState: boolean;
  setDarkState: React.Dispatch<React.SetStateAction<boolean>>;
  theme: Theme;
  favoriteVideos: FavoriteVideos;
  dispatchFav: React.Dispatch<Action>;
  authenticated: boolean;
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
  avatar: string;
  setAvatar: React.Dispatch<React.SetStateAction<string>>;
}

interface GlobalContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const GlobalContext = React.createContext<GlobalContextValues>({
  search: '',
  setSearch: () => { return; },
  darkState: false,
  setDarkState: () => { return; },
  theme: createMuiTheme({
    palette: {
      type: 'light',
    }
  }),
  favoriteVideos: {},
  dispatchFav: () => { return; },
  authenticated: false,
  setAuthenticated: () => { return; },
  avatar: '',
  setAvatar: () => { return; },
});

const GlobalContextProvider: React.FunctionComponent<GlobalContextProviderProps> =
  (props: GlobalContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const [darkState, setDarkState] = useState<boolean>(false);
    const [theme, setTheme] = useState<Theme>(
      createMuiTheme({
        palette: {
          type: 'light',
        },
      })
    );
    const [favoriteVideos, dispatchFav] = useReducer(reducer, initialState);
    const [authenticated, setAuthenticated] = useState<boolean>(false);
    const [avatar, setAvatar] = useState<string>('');
    const { user, isAuthenticated } = useAuth0();

    useEffect(
      () => {
        setAuthenticated(isAuthenticated);
      },
      [isAuthenticated]
    );

    useEffect(
      () => {
        if (user) {
          setAvatar(user.picture);
        }
      },
      [user]
    );

    useEffect(
      () => {
        setAuthenticated(Boolean(storage.get(AUTH_STORAGE_KEY)));
        setAvatar(String(storage.get(AUTH_AVATAR)));
        dispatchFav({
          type: 'addAll',
          all: JSON.parse(String(storage.get(FAVORITES_VIDEOS))),
        });
      },
      []
    );

    useEffect(
      () => {
        const palletType = darkState ? 'dark' : 'light';
        setTheme(
          createMuiTheme({
            palette: {
              type: palletType,
            }
          })
        );
      },
      [darkState]
    );

    useEffect(
      () => {
        storage.set(FAVORITES_VIDEOS, JSON.stringify(favoriteVideos));
      },
      [favoriteVideos]
    );

    const value: GlobalContextValues = {
      search,
      setSearch,
      darkState,
      setDarkState,
      theme,
      favoriteVideos,
      dispatchFav,
      authenticated,
      setAuthenticated,
      avatar,
      setAvatar,
    };

    return (
      <NoSsr>
        <MuiThemeProvider theme={theme}>
          <ThemeProvider theme={theme}>
            <GlobalContext.Provider
              value={value}
            >
              {props.children}
            </GlobalContext.Provider>
          </ThemeProvider>
        </MuiThemeProvider>
      </NoSsr>
    );
  };

export {GlobalContextProvider, GlobalContext};