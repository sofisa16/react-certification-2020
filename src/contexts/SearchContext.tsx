import React, {useState} from 'react';

interface SearchContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
}

interface SearchContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const SearchContext = React.createContext<SearchContextValues>({
  search: "",
  setSearch: () => { return; },
});

const SearchContextProvider: React.FunctionComponent<SearchContextProviderProps> =
  (props: SearchContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const value: SearchContextValues = {
      search,
      setSearch,
    };

    return (
      <SearchContext.Provider
        value={value}
      >
        {props.children}
      </SearchContext.Provider>
    );
  };

export {SearchContextProvider, SearchContext};