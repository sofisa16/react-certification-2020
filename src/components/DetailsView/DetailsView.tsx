import React, {useState, useEffect, useContext} from 'react';
import {useParams} from 'react-router-dom';
import styled from 'styled-components';
import useYouTubeAPI from '../../hooks/useYouTubeAPI';
import {YouTubeResponse, YouTubeResponseItems} from '../../data-types/YoutubeAPI';
//import videos_id from './../../data/videos_id.json';
//import bug from './../../data/bug.json';
//import related_videos from './../../data/related_videos.json';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import RelatedVideo from '../RelatedVideo/RelatedVideo';
import {GlobalContext} from '../../contexts/GlobalContext';

export interface DetailsViewParams {
  videoId: string;
}

const ParentContainer = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: row;
  padding: 20px;
  box-sizing: border-box;
`;

const RightContainer = styled.div`
  width: 70%;
  height: 100%;
  position: relative;
  display: flex;
  flex-direction: column;
  padding-right: 15px;
`;

const LeftContainer = styled.div`
  width: 30%;
  height: 100%;
  overflow: scroll;
`;

const DownContainer = styled.div`
  padding: 15px 30px;
`;

const TitleContainer = styled.div`
  padding-bottom: 15px;
  display: grid;
  grid-template-columns: 1fr auto;
`;

const JustifyText = styled(Typography)`
  text-align: justify;
`;

const Iframe = styled.div`
  > iframe {
    width: 100%;
    height: 500px;
  }
`;

interface DetailsViewProps {
  relatedItems: YouTubeResponseItems[],
}

function DetailsView(props: DetailsViewProps): JSX.Element {
  const {relatedItems} = props;
  const {videoId} = useParams<DetailsViewParams>();
  const {getFromYouTubeAPI} = useYouTubeAPI();
  const [items, setItems] = useState<YouTubeResponseItems[]>([]);
  const [embedHtml, setEmbedHtml] = useState<string>('');
  const [relatedVideos, setRelatedVideos] = useState<JSX.Element[]>([]);
  const {favoriteVideos, dispatchFav, authenticated} = useContext(GlobalContext);
  const [buttonLabel, setButtonLabel] = useState<string>('Agregar a favoritos');

  useEffect(
    () => {
      async function getVideoFromYoutube(videoId: string): Promise<void> {
        try {
          const response: Response = await getFromYouTubeAPI(`videos?part=contentDetails&part=id&part=player&part=snippet&part=statistics&id=${videoId}`);
          const result: YouTubeResponse = await response.json();
          
          setItems(result.items);
        }
        catch (error) {
          console.log(error);
        }
      }

      getVideoFromYoutube(videoId);
    },
    [videoId]
  );

  useEffect(
    () => {
      if(items[0]?.player?.embedHtml) {
        setEmbedHtml(items[0].player.embedHtml);
      }
    },
    [items]
  );

  useEffect(
    () => {
      const related: JSX.Element[] = [];
      for(const item of relatedItems) {
        const id = typeof(item.id) === 'string' ? item.id : item?.id?.videoId;
        if(item.snippet) {
          related.push(
            <RelatedVideo
              title={item.snippet.title}
              thumbnails={item.snippet.thumbnails.default.url}
              videoId={`${id}`}
              key={`${item.snippet.title}${item.snippet.publishedAt}`}
            />
          );
        }
      }
      setRelatedVideos(related);
    },
    [relatedItems]
  );

  function iframe(embedHtml: string): {__html: string;} {
    return {
      __html: embedHtml,
    };
  }

  function toogleAddFavorite(): void {
    const id = typeof(items[0].id) === 'string' ? items[0].id : items[0]?.id?.videoId;
    if (favoriteVideos[`${id}`] && Object.keys(favoriteVideos[`${id}`]).length !== 0) {
      dispatchFav({
        type: 'remove',
        payload: items[0],
      });
    }
    else {
      dispatchFav({
        type: 'add',
        payload: items[0],
      });
    }
  }

  useEffect(
    () => {
      const id = items && items[0]
        ? typeof(items[0].id) === 'string'
          ? items[0].id
          : items[0]?.id?.videoId
        : undefined;
      if (favoriteVideos[`${id}`] && Object.keys(favoriteVideos[`${id}`]).length !== 0) {
        setButtonLabel('Remover de favoritos');
      }
      else {
        setButtonLabel('Agregar a favoritos');
      }
    },
    [favoriteVideos, videoId, items]
  );

  return (
    <ParentContainer>
      <RightContainer>
        <Iframe dangerouslySetInnerHTML={iframe(embedHtml)}></Iframe>
        <DownContainer>
          <TitleContainer>
            <Typography gutterBottom variant='h5' component='h2'>
              {items[0]?.snippet?.title}
            </Typography>
            {
              authenticated &&
              <Button onClick={toogleAddFavorite}>{buttonLabel}</Button>
            }
          </TitleContainer>
          <JustifyText variant='body2' color='textSecondary'>
            {items[0]?.snippet?.description}
          </JustifyText>
        </DownContainer>
      </RightContainer>
      <LeftContainer>
        {relatedVideos}
      </LeftContainer>
    </ParentContainer>
  );
}

export default DetailsView;