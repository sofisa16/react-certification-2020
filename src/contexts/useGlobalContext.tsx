import {useContext} from 'react';
import {GlobalContext, GlobalContextValues} from './GlobalContext';

function useGlobalContext(): GlobalContextValues {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Please use the "useContext" hook with a GlobalContextProvider.')
  }

  return context;
}

export default useGlobalContext;