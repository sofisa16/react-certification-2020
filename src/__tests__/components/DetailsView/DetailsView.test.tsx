import React from 'react';
import { render } from '@testing-library/react';
import DetailsView from '../../../components/DetailsView/DetailsView';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import {GlobalContextProvider} from '../../../contexts/GlobalContext';
import CssBaseline from '@material-ui/core/CssBaseline';
import related_videos from '../../../data/related_videos.json';

it('renders DetailsView component', () => {
  const history = createMemoryHistory();
  const route = '/favorites/fLclGPr7fj4';
  history.push(route);

  const tree = render(
    <GlobalContextProvider>
      <CssBaseline>
        <Router history={history}>
          <DetailsView
            relatedItems={related_videos.items}
            path='favorites/'
          />
        </Router>
      </CssBaseline>
    </GlobalContextProvider>
  );
  expect(tree).toMatchSnapshot();
});