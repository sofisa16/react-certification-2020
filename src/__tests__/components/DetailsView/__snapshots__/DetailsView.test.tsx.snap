// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DetailsView component 1`] = `null`;
